packages/tableofcontents.lua

```diff
@@ -159,8 +92,96 @@ SILE.doTexlike([[%
 \define[command=tableofcontents:level2number]{}%
 \define[command=tableofcontents:level3number]{}%
 ]])
+end)
+
+end
 
-  end,
+local function registerCommands (class)
+
+  SILE.registerCommand("tableofcontents", function (options, _)
+    local depth = SU.cast("integer", options.depth or 3)
+    local linking = SU.boolean(options.linking, true)
+    local tocfile,_ = io.open(SILE.masterFilename .. '.toc')
+    if not tocfile then
+      SILE.call("tableofcontents:notocmessage")
+      return
+    end
+    local doc = tocfile:read("*all")
+    local toc = assert(load(doc))()
+    SILE.call("tableofcontents:header")
+    for i = 1, #toc do
+      local item = toc[i]
+      if item.level <= depth then
+        SILE.call("tableofcontents:item", {
+          level = item.level,
+          pageno = item.pageno,
+          number = item.number,
+          link = linking and item.link
+        }, item.label)
+      end
+    end
+    SILE.call("tableofcontents:footer")
+    _tableofcontents = toc
+  end)
+
+  SILE.registerCommand("tableofcontents:item", function (options, content)
+    SILE.settings:temporarily(function ()
+      SILE.settings:set("typesetter.parfillskip", SILE.nodefactory.glue())
+      SILE.call("tableofcontents:level" .. options.level .. "item", {
+      }, _linkWrapper(options.link,
+      function ()
+        SILE.call("tableofcontents:level" .. options.level .. "number", {
+        }, function ()
+          if options.number then
+            SILE.typesetter:typeset(options.number or "")
+            SILE.call("kern", { width = "1spc" })
+          end
+        end)
+        SILE.process(content)
+        SILE.call("dotfill")
+        SILE.typesetter:typeset(options.pageno)
+      end)
+      )
+    end)
+  end)
+
+  SILE.registerCommand("tocentry", function (options, content)
+    local dest
+    if SILE.Commands["pdf:destination"] then
+      dest = "dest" .. dc
+      SILE.call("pdf:destination", { name = dest })
+      SILE.typesetter:pushState()
+      SILE.process(content)
+      local title = _nodesToText(SILE.typesetter.state.nodes)
+      SILE.typesetter:popState()
+      SILE.call("pdf:bookmark", { title = title, dest = dest, level = options.level })
+      dc = dc + 1
+    end
+    SILE.call("info", {
+      category = "toc",
+      value = {
+        label = content,
+        level = (options.level or 1),
+        number = options.number,
+        link = dest
+      }
+    })
+  end)
+
+  SILE.registerCommand("tableofcontents:title", function (_, _)
+    SU.deprecated("\\tableofcontents:title", "\\fluent{toc-title}", "0.13.0", "0.14.0")
+    SILE.call("fluent", {}, { "toc-title" })
+  end)
+
+end
+
+return {
+  init = init,
+  registerCommands = registerCommands,
+  exports = {
+    writeToc = writeToc,
```

Comment:
Why do we need to still export these functions (and pollute the class with them), if they are to be called via the new hooks we registered above?

Reply:
Good point, maybe we don't...

Unless it's possible some other package would need to call it directly.